Source/WebCore/page/DragController.cpp

```diff
@@ -239,74 +239,102 @@ inline static bool dragIsHandledByDocument(DragHandlingMethod dragHandlingMethod
+    if (frame.contentRenderer()) {
+        static constexpr OptionSet<HitTestRequest::Type> hitType { HitTestRequest::Type::ReadOnly, HitTestRequest::Type::Active, HitTestRequest::Type::DisallowUserAgentShadowContent, HitTestRequest::Type::AllowChildFrameContent };
+        hitTestResult = frame.eventHandler().hitTestResultAtPoint(point, hitType);
+    }
+    m_documentUnderMouse = hitTestResult.innerNode() ? &hitTestResult.innerNode()->document() : nullptr;
 
     disallowFileAccessIfNeeded(dragData);
 
     ShouldOpenExternalURLsPolicy shouldOpenExternalURLsPolicy = ShouldOpenExternalURLsPolicy::ShouldNotAllow;
     if (RefPtr document = m_documentUnderMouse)
         shouldOpenExternalURLsPolicy = document->shouldOpenExternalURLsPolicyToPropagate();
 
-    if ((m_dragDestinationActionMask.contains(DragDestinationAction::DHTML)) && dragIsHandledByDocument(m_dragHandlingMethod)) {
-        client().willPerformDragDestinationAction(DragDestinationAction::DHTML, dragData);
-        bool preventedDefault = false;
-        if (localMainFrame->view())
-            preventedDefault = localMainFrame->eventHandler().performDragAndDrop(createMouseEvent(dragData), Pasteboard::create(dragData), dragData.draggingSourceOperationMask(), dragData.containsFiles());
+    auto continuePerformDrag = [weakThis = WeakPtr { *this }, this, completionHandler = WTFMove(completionHandler), dragData, weakFrame = WeakPtr { frame }, shouldOpenExternalURLsPolicy](bool preventedDefault) mutable {
+        if (!weakThis || !weakFrame) {
+            completionHandler(false);
+            return;
+        }
         if (preventedDefault) {
             clearDragCaret();
             m_documentUnderMouse = nullptr;
-            return true;
+            completionHandler(true);
+            return;
+        }
+        if ((m_dragDestinationActionMask.contains(DragDestinationAction::Edit)) && concludeEditDrag(dragData)) {
+            client().didConcludeEditDrag();
+            m_documentUnderMouse = nullptr;
+            clearDragCaret();
+            completionHandler(true);
+            return;
         }
-    }
 
-    if ((m_dragDestinationActionMask.contains(DragDestinationAction::Edit)) && concludeEditDrag(dragData)) {
-        client().didConcludeEditDrag();
         m_documentUnderMouse = nullptr;
         clearDragCaret();
-        return true;
-    }
 
-    m_documentUnderMouse = nullptr;
-    clearDragCaret();
-
-    if (!operationForLoad(dragData))
-        return false;
+        if (!operationForLoad(dragData)) {
+            completionHandler(false);
+            return;
+        }
+        auto frameID = weakFrame->frameID();
+        client().willPerformDragDestinationAction(DragDestinationAction::Load, dragData, frameID, [weakThis = WTFMove(weakThis), completionHandler = WTFMove(completionHandler), weakFrame = WTFMove(weakFrame), shouldOpenExternalURLsPolicy, dragData] mutable {
+            if (!weakThis) {
```

Comment:
I think you're missing a weakFrame check here.  You dereference it without checking if it's still non-null after willPerformDragDestinationAction has completed.

Also, it would be better to get a RefPtr out of it and check that, so you have something on the stack ensuring the object won't get destroyed.